benchmark/common.js

```diff
@@ -1,205 +1,207 @@
+  const match = arg.split('=');
+  const isNumber = isFinite(match[1]) && match[1] !== '';
+  configArgs[match[0]] = isNumber ? parseFloat(match[1]) : match[1];
 }
 
+// The port used by servers and wrk
+exports.PORT = process.env.PORT || 12346;
+
 exports.createBenchmark = function(fn, options) {
   return new Benchmark(fn, options);
 };
 
 function Benchmark(fn, options) {
   this.fn = fn;
-  this.options = options;
-  this.config = parseOpts(options);
-  this._name = require.main.filename.split(/benchmark[\/\\]/).pop();
-  this._start = [0, 0];
+  this.name = require.main.filename.slice(__dirname.length + 1);
+  this.queue = this._queue(options);
+  this.config = this.queue[0];
+
+  this._time = [0, 0]; // keep process.hrtime values
   this._started = false;
 
-  var self = this;
+  process.nextTick(() => this._run());
+}
 
-  process.nextTick(function() {
-    self._run();
-  });
+Benchmark.prototype._queue = function(options) {
+  const queue = [];
+  const keys = Object.keys(options);
+
+  // Overwrite options with configuration specified in the process arguments
+  options = Object.assign({}, options);
+  for (const key of Object.keys(configArgs)) {
+    options[key] = [ configArgs[key] ];
+  }
+
+  // Perform a breadth-first walk though all options to genereate a
+  // configuration list that contains all combinations.
+  function recursive(keyIndex, prevConfig) {
+    const key = keys[keyIndex];
+    const values = options[key];
+
+    for (const value of values) {
+      if (typeof value !== 'number' && typeof value !== 'string') {
+        throw new TypeError(`configuration "${key}" had type ${typeof value}`);
+      }
+
+      const currConfig = Object.assign({ [key]: value }, prevConfig);
+
+      if (keyIndex + 1 < keys.length) {
+        recursive(keyIndex + 1, currConfig);
+      } else {
+        queue.push(currConfig);
+      }
+    }
+  }
+
+  if (keys.length > 0) {
+    recursive(0, {});
+  } else {
+    queue.push({});
+  }
+
+  return queue;
+};
+
+function hasWrk() {
+  const result = child_process.spawnSync('wrk', ['-h']);
+  if (result.error && result.error.code === 'ENOENT') {
+    console.error('Couldn\'t locate `wrk` which is needed for running ' +
+      'benchmarks. Check benchmark/README.md for further instructions.');
+    process.exit(1);
+  }
 }
 
 // benchmark an http server.
-Benchmark.prototype.http = function(p, args, cb) {
+const WRK_REGEXP = /Requests\/sec:[ \t]+([0-9\.]+)/;
+Benchmark.prototype.http = function(urlPath, args, cb) {
   hasWrk();
-  var self = this;
-  var regexp = /Requests\/sec:[ \t]+([0-9\.]+)/;
-  var url = 'http://127.0.0.1:' + exports.PORT + p;
+  const self = this;
 
-  args = args.concat(url);
+  const urlFull = 'http://127.0.0.1:' + exports.PORT + urlPath;
+  args = args.concat(urlFull);
 
-  var out = '';
-  var child = child_process.spawn('wrk', args);
+  const childStart = process.hrtime();
+  const child = child_process.spawn('wrk', args);
```

Comment:
The benchmark don't directly use `process.hrtime()` in the http case, but instead the op/sec reported by `wrk` in the stdout. The `childStart` stuff is just for reporting how much time the benchmark actually use. This is only used as a side metic, which I found useful for finding benchmarks that uses to many iterations.